src/games/strategy/triplea/delegate/MoveValidator.java

```diff
@@ -1409,47 +1422,58 @@ public static String validateCanal(final Route route, final Collection<Unit> uni
    */
   public static String validateCanal(final Territory territory, final Route route, final Collection<Unit> units,
       final PlayerID player, final GameData data) {
+    String failureMessage = null;
     final Set<CanalAttachment> canalAttachments = CanalAttachment.get(territory);
-    if (canalAttachments.isEmpty()) {
-      return null;
-    }
-    final Iterator<CanalAttachment> iter = canalAttachments.iterator();
-    while (iter.hasNext()) {
-      final CanalAttachment attachment = iter.next();
-      if (attachment == null) {
+    for (final CanalAttachment canalAttachment : canalAttachments) {
+      if (!needToCheckCanal(canalAttachment, route, data)) {
         continue;
       }
-      if (route != null) {
-        boolean mustCheck = false;
-        Territory last = null;
-        final Set<Territory> connectionToCheck = CanalAttachment.getAllCanalSeaZones(attachment.getCanalName(), data);
-        for (final Territory current : route.getAllTerritories()) {
-          if (last != null) {
-            final Collection<Territory> lastTwo = new ArrayList<Territory>();
-            lastTwo.add(last);
-            lastTwo.add(current);
-            mustCheck = lastTwo.containsAll(connectionToCheck);
-            if (mustCheck) {
-              break;
-            }
-          }
-          last = current;
-        }
-        if (!mustCheck) {
-          continue;
+      failureMessage = canPassThroughCanal(canalAttachment, units, player, data);
+      if (failureMessage == null) {
+        return null;
+      }
+    }
+    return failureMessage;
+  }
+
+  /**
+   * Checks if route is either null or includes both canal territories so needs to be checked.
+   */
+  private static boolean needToCheckCanal(final CanalAttachment canalAttachment, final Route route, final GameData data) {
+    if (route == null) {
+      return true;
+    }
+    Territory last = null;
+    final Set<Territory> connectionToCheck = CanalAttachment.getAllCanalSeaZones(canalAttachment.getCanalName(), data);
+    for (final Territory current : route.getAllTerritories()) {
+      if (last != null) {
+        final Collection<Territory> lastTwo = new ArrayList<Territory>();
+        lastTwo.add(last);
+        lastTwo.add(current);
+        if (lastTwo.containsAll(connectionToCheck)) {
+          return true;
         }
       }
-      if (units != null && Match.allMatch(units, Matches.unitIsOfTypes(attachment.getExcludedUnits(data)))) {
-        continue;
+      last = current;
+    }
+    return false;
+  }
+
+  /**
+   * Checks if units can pass through canal and returns null if true or a failure message if false.
+   */
+  private static String canPassThroughCanal(final CanalAttachment canalAttachment, final Collection<Unit> units,
```

Comment:
Even better yet, I think this method should have been broken up a bit differently/further. We also have a SRP violation since this method checks many different things. I recommend this be fixed further by splitting this up into two or three methods that return simple "true/false" answers. Let the caller code then formulate the message that is displayed.